doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,271 @@
+    called again **once**.
+
+The following use cases and behaviors should be implemented.
+
+ - The API and ReactAL should be robust. Since ReactAL is working concurrently a 
+    wrong usage could lead to race conditions, extremely hard to debug bug reports
+    (Heisenbugs) and can be frustrating to the user.
+    We list here features marked with [robust] which are only supported to 
+    increase this kind of the robustness.
+
+ - concurrently: assigning a callback at any time from anywhere. This means in particular
+    - Attaching a callback from within a callback.
+    - [robust] Updating a callback from within the same callback 
+      ```cpp
+      void onSampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived2);
+      }
+
+      myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+      ```
+ - One can assign at most one callback to a specific event of an object. The event is usually defined 
+    with an enum by the developer. One example is `SubscriberEvent::HAS_SAMPLE_RECEIVED`.
+    - Attaching a callback to an event where a callback has been already assigned overrides
+      the existing callback with the new one.
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished. After the successful removal the 
+                 callback will continue and will not be called again.
+    - Calling detach means that after that call the callback is no longer assigned
+      even when it was not assigned in the first place. Therefore it will always succeed.
+
+ - When the ReactAL goes out of scope it should detach itself from every class 
+     to which it was assigned via a callback provided by the attached class (like in the WaitSet).
+
+ - When the class which is assigned to the ReactAL goes out of scope it should 
+    detach itself from the ReactAL via a callback provided by the ReactAL (like in the WaitSet).
+
+### Usage Code Example
+```cpp
+ReactAL myCallbackReactAL;
+iox::popo::UntypedSubscriber mySubscriber;
+
+void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+  subscriber.take().and_then([&](auto & sample){
+    std::cout << "received " << std::hex << sample->payload() << std::endl;
+  });
+}
+
+void onWaitForSubscription(iox::popo::UntypedSubscriber & subscriber) {
```

Comment:
I think the name is wrong. This isn't executed while waiting for subscription, but on subscription.